File: src/YowThi.Erp.Application/Party/PartyLifecycleOptions.cs
namespace YowThi.Erp.Application.Party;

public sealed record PartyLifecycleOptionsQuery(
    string Locale,
    string? Search,
    int Offset,
    int Limit);

public sealed record PartyLifecycleOptionPage<T>(
    IReadOnlyList<T> Items,
    int? NextOffset);

public sealed record PartyLifecycleOption(
    Guid Id,
    string DisplayName,
    bool Active,
    long RowVersion,
    bool Deleted,
    DateTimeOffset? DeletedAt);

public interface IPartyLifecycleOptionsReader
{
    ValueTask<PartyLifecycleOptionPage<PartyLifecycleOption>> GetSuppliersAsync(
        PartyLifecycleOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<PartyLifecycleOptionPage<PartyLifecycleOption>> GetCustomersAsync(
        PartyLifecycleOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<PartyLifecycleOptionPage<PartyLifecycleOption>> GetOutsourcedVendorsAsync(
        PartyLifecycleOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<PartyLifecycleOptionPage<PartyLifecycleOption>> GetFarmersAsync(
        PartyLifecycleOptionsQuery query,
        CancellationToken cancellationToken);

    ValueTask<PartyLifecycleOptionPage<PartyLifecycleOption>> GetEmployeesAsync(
        PartyLifecycleOptionsQuery query,
        CancellationToken cancellationToken);
}

File: src/YowThi.Erp.Web/src/features/party/partyLifecycleOptions.ts
import { getApiJson } from '../../app/api/apiTransport';
import type { OperationalLocale } from '../../app/i18n/locale';

export type PartyLifecycleKind =
  | 'suppliers'
  | 'customers'
  | 'outsourced-vendors'
  | 'farmers'
  | 'employees';

export interface PartyLifecycleOption {
  id: string;
  displayName: string;
  active: boolean;
  rowVersion: number;
  deleted: boolean;
  deletedAt: string | null;
}

interface PartyLifecycleOptionPage {
  items: PartyLifecycleOption[];
  nextCursor: string | null;
}

interface PartyLifecycleOptionsQuery {
  locale: OperationalLocale;
  search?: string;
  cursor?: string | null;
  limit?: number;
  signal?: AbortSignal;
}

export function listPartyLifecycleOptions(
  kind: PartyLifecycleKind,
  query: PartyLifecycleOptionsQuery,
): Promise<PartyLifecycleOptionPage> {
  const parameters = new URLSearchParams();
  const search = query.search?.trim();
  if (search) parameters.set('search', search);
  if (query.cursor) parameters.set('cursor', query.cursor);
  if (query.limit !== undefined) parameters.set('limit', String(query.limit));
  const suffix = parameters.size === 0 ? '' : `?${parameters.toString()}`;

  return getApiJson<PartyLifecycleOptionPage>(
    `/api/v1/party/lifecycle-options/${kind}${suffix}`,
    { locale: query.locale, signal: query.signal },
  );
}
